File: src/main/java/ru/timeconqueror/timecore/common/capability/property/serializer/IPropertySerializer.java
package ru.timeconqueror.timecore.common.capability.property.serializer;

import net.minecraft.nbt.CompoundTag;

public interface IPropertySerializer<T> {
    void serialize(String name, T value, CompoundTag nbt);
    T deserialize(String name, CompoundTag nbt);

    static <T> IPropertySerializer<T> nullable(IPropertySerializer<T> serializer) {
        return new NullPropertySerializer<>(serializer);
    }
}

File: src/main/java/ru/timeconqueror/timecore/common/capability/property/serializer/IPropertySerializer.kt
package ru.timeconqueror.timecore.common.capability.property.serializer

import net.minecraft.nbt.CompoundTag


interface IPropertySerializer<T> {

    fun serialize(name: String, value: T, nbt: CompoundTag)

    fun deserialize(name: String, nbt: CompoundTag): T

    fun nullable() = NullPropertySerializer(this)
}